File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/manager/controller/MediaController.java
package com.trtc.uikit.livekit.manager.controller;

import static com.tencent.liteav.beauty.TXBeautyManager.TXBeautyStyleSmooth;

import android.content.Context;
import android.view.View;

import com.tencent.cloud.tuikit.engine.common.TUICommonDefine;
import com.tencent.cloud.tuikit.engine.common.TUIVideoView;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.tencent.qcloud.tuicore.TUIConfig;
import com.tencent.qcloud.tuicore.permission.PermissionCallback;
import com.trtc.tuikit.common.livedata.Observer;
import com.trtc.uikit.livekit.common.uicomponent.beauty.TEBeautyService;
import com.trtc.uikit.livekit.common.utils.LiveKitLog;
import com.trtc.uikit.livekit.common.utils.PermissionRequest;
import com.trtc.uikit.livekit.manager.error.ErrorHandler;
import com.trtc.uikit.livekit.service.ILiveService;
import com.trtc.uikit.livekit.state.LiveDefine;
import com.trtc.uikit.livekit.state.LiveState;

public class MediaController extends Controller {
    private static final String TAG = "MediaController";

    private final Observer<LiveDefine.LinkStatus> mLinkStateObserver = this::onLinkStateChanged;
    private final TEBeautyService                 mTEBeautyService;

    public MediaController(LiveState state, ILiveService service) {
        super(state, service);
        mViewState.linkStatus.observe(mLinkStateObserver);
        mTEBeautyService = new TEBeautyService(state, service.getTRTCCloud());
    }

    @Override
    public void destroy() {
        LiveKitLog.info(TAG + " destroy");
        mViewState.linkStatus.removeObserver(mLinkStateObserver);
        mTEBeautyService.clearBeautyView();
        closeLocalCamera();
    }

    public void operateMicrophone() {
        boolean isMicrophoneOpened = mMediaState.isMicrophoneOpened.get();
        if (!isMicrophoneOpened) {
            openLocalMicrophone();
            return;
        }
        boolean isMuted = mMediaState.isMicrophoneMuted.get();
        if (isMuted) {
            unMuteLocalAudio();
        } else {
            muteLocalAudio();
        }
    }

    public void setLocalVideoView(TUIVideoView view) {
        mLiveService.setLocalVideoView(view);
    }

    public void setRemoteVideoView(String userId, TUIRoomDefine.VideoStreamType streamType, TUIVideoView videoView) {
        mLiveService.setRemoteVideoView(userId, streamType, videoView);
    }

    public void startPlayRemoteVideo(String userId, TUIRoomDefine.VideoStreamType streamType,
                                     TUIRoomDefine.PlayCallback callback) {
        mLiveService.startPlayRemoteVideo(userId, streamType, callback);
    }

    public void openLocalCamera() {
        boolean hasCameraPermission = mMediaState.hasCameraPermission.get();
        if (hasCameraPermission) {
            openLocalCameraByService();
            return;
        }
        LiveKitLog.info(TAG + " requestPermissions:[]");
        PermissionRequest.requestPermissions(TUIConfig.getAppContext(), new PermissionCallback() {
            @Override
            public void onRequesting() {
                LiveKitLog.info(TAG + " requestPermissions:[onRequesting]");
            }

            @Override
            public void onGranted() {
                LiveKitLog.info(TAG + " requestPermissions:[onGranted]");
                mMediaState.hasCameraPermission.set(true);
                openLocalCameraByService();
            }

            @Override
            public void onDenied() {
                LiveKitLog.error(TAG + " requestPermissions:[onDenied]");
            }
        });
    }

    public void closeLocalCamera() {
        mLiveService.closeLocalCamera();
        mMediaState.isCameraOpened.set(false);
    }

    public void switchCamera() {
        boolean isFrontCamera = mMediaState.isFrontCamera.get();
        mLiveService.switchCamera(!isFrontCamera);
        mMediaState.isFrontCamera.set(!isFrontCamera);
    }

    public void setCameraMirror() {
        boolean isMirror = mMediaState.isMirror.get();
        mLiveService.setCameraMirror(!isMirror);
        mMediaState.isMirror.set(!isMirror);
    }

    public void updateVideoQuality(TUIRoomDefine.VideoQuality quality) {
        mLiveService.updateVideoQuality(quality);
        mMediaState.videoQuality.set(quality);
    }

    public void updateAudioQuality(TUIRoomDefine.AudioQuality quality) {
        mLiveService.updateAudioQuality(quality);
        mMediaState.audioQuality.set(quality);
    }

    public void setBeautyLevel(int level) {
        mLiveService.setBeautyLevel(level);
        mBeautyState.smoothLevel.set(level);
    }

    public void setWhitenessLevel(int level) {
        mLiveService.setWhitenessLevel(level);
        mBeautyState.whitenessLevel.set(level);
    }

    public void setRuddyLevel(int level) {
        mLiveService.setRuddyLevel(level);
        mBeautyState.ruddyLevel.set(level);
    }

    public void closeBeauty() {
        mBeautyState.smoothLevel.set(0);
        mLiveService.setBeautyLevel(0);
        mBeautyState.whitenessLevel.set(0);
        mLiveService.setWhitenessLevel(0);
        mBeautyState.ruddyLevel.set(0);
        mLiveService.setRuddyLevel(0);
    }

    public void setCustomVideoProcess() {
        mTEBeautyService.setCustomVideoProcess();
    }

    public View getTEBeautyView(Context context) {
        return mTEBeautyService.getBeautyView(context);
    }

    private void unMuteLocalAudio() {
        mLiveService.unMuteLocalAudio(new TUIRoomDefine.ActionCallback() {
            @Override
            public void onSuccess() {
                mMediaState.isMicrophoneMuted.set(false);
            }

            @Override
            public void onError(TUICommonDefine.Error error, String message) {
                ErrorHandler.onError(error);
            }
        });
    }

    private void muteLocalAudio() {
        mLiveService.muteLocalAudio();
        mMediaState.isMicrophoneMuted.set(true);
    }

    private void openLocalMicrophone() {
        boolean hasMicrophonePermission = mMediaState.hasMicrophonePermission.get();
        if (hasMicrophonePermission) {
            openLocalMicrophoneByService();
            return;
        }
        LiveKitLog.info(TAG + " requestMicrophonePermissions:[]");
        PermissionRequest.requestMicrophonePermissions(TUIConfig.getAppContext(), new PermissionCallback() {
            @Override
            public void onRequesting() {
                LiveKitLog.info(TAG + " requestMicrophonePermissions:[onRequesting}");
            }

            @Override
            public void onGranted() {
                LiveKitLog.info(TAG + " requestMicrophonePermissions:[onGranted]");
                mMediaState.hasMicrophonePermission.set(true);
                openLocalMicrophoneByService();
            }

            @Override
            public void onDenied() {
                LiveKitLog.warn(TAG + " requestMicrophonePermissions:[onDenied]");
            }
        });
    }

    private void openLocalMicrophoneByService() {
        unMuteLocalAudio();
        mLiveService.openLocalMicrophone(new TUIRoomDefine.ActionCallback() {
            @Override
            public void onSuccess() {
                mMediaState.isMicrophoneOpened.set(true);
            }

            @Override
            public void onError(TUICommonDefine.Error error, String message) {
                ErrorHandler.onError(error);
            }
        });
    }

    private void openLocalCameraByService() {
        if (mMediaState.isCameraOpened.get()) {
            LiveKitLog.error(TAG + " camera is opened, no need to open again");
            return;
        }
        mTEBeautyService.initBeautyKit();
        boolean isFront = mMediaState.isFrontCamera.get();
        TUIRoomDefine.VideoQuality quality = mMediaState.videoQuality.get();
        mLiveService.openLocalCamera(isFront, quality, new TUIRoomDefine.ActionCallback() {
            @Override
            public void onSuccess() {
                initLivingConfig();
                mMediaState.isCameraOpened.set(true);
            }

            @Override
            public void onError(TUICommonDefine.Error error, String message) {
            }
        });
    }

    private void initLivingConfig() {
        mLiveService.enableGravitySensor(true);
        mLiveService.setVideoResolutionMode(TUIRoomDefine.ResolutionMode.PORTRAIT);
        mLiveService.setBeautyStyle(TXBeautyStyleSmooth);
        updateVideoQuality(mMediaState.videoQuality.get());
        setBeautyLevel(mBeautyState.smoothLevel.get());
        setWhitenessLevel(mBeautyState.whitenessLevel.get());
        setRuddyLevel(mBeautyState.ruddyLevel.get());
        updateAudioQuality(mMediaState.audioQuality.get());
    }

    private void onLinkStateChanged(LiveDefine.LinkStatus status) {
        LiveKitLog.info(TAG + " onUserStateChanged status:" + status);
        if (status == LiveDefine.LinkStatus.LINKING) {

            if (mViewState.autoOpenCameraOnSeated.get()) {
                LiveKitLog.info(TAG + " requestPermissions:[Camera, Microphone]}");
                PermissionRequest.requestPermissions(TUIConfig.getAppContext(), new PermissionCallback() {
                    @Override
                    public void onRequesting() {
                        LiveKitLog.info(TAG + " requestPermissions:[Camera, Microphone]:[onRequesting}");
                    }

                    @Override
                    public void onGranted() {
                        LiveKitLog.info(TAG + " requestPermissions:[Camera, Microphone]:[onGranted]");
                        mMediaState.hasMicrophonePermission.set(true);
                        mMediaState.hasCameraPermission.set(true);
                        openLocalMicrophone();
                        openLocalCamera();
                    }

                    @Override
                    public void onDenied() {
                        LiveKitLog.warn(TAG + " requestPermissions:[Camera, Microphone]:[onDenied]");
                    }

                });
            } else {
                openLocalMicrophone();
            }
        }
    }
}

File: Flutter/lib/manager/controller/media_controller.dart
import 'dart:ui';

import 'package:permission_handler/permission_handler.dart';
import 'package:rtc_room_engine/api/common/tui_common_define.dart';
import 'package:rtc_room_engine/api/room/tui_room_define.dart';
import 'package:tencent_live_uikit/common/index.dart';
import 'package:tencent_live_uikit/manager/controller/controller.dart';
import 'package:tencent_live_uikit/state/index.dart';

class MediaController extends Controller {
  static const String tag = 'MediaController';
  late VoidCallback linkStateObserver = () => _onLinkStateChanged();

  MediaController(super.state, super.service) {
    viewState.linkStatus.addListener(linkStateObserver);
  }

  @override
  void destroy() {
    viewState.linkStatus.removeListener(linkStateObserver);
  }

  void operateMicrophone() {
    final isMicrophoneOpened = mediaState.isMicrophoneOpened.value;
    if (!isMicrophoneOpened) {
      unMuteLocalAudio();
      openLocalMicrophone();
      return;
    }
    final isMuted = mediaState.isMicrophoneMuted.value;
    if (isMuted) {
      unMuteLocalAudio();
    } else {
      muteLocalAudio();
    }
  }

  Future<TUIActionCallback> openLocalMicrophone() async {
    var microphone = await Permission.microphone.request();

    if (!microphone.isGranted) {
      LiveKitLogger.error('requestMicrophonePermissions:[onDenied]');
      return TUIActionCallback(code: TUIError.errPermissionDenied, message: "camera permission denied");
    }
    unMuteLocalAudio();
    var result = await liveService.openLocalMicrophone(TUIAudioQuality.audioProfileDefault);
    if (result.code == TUIError.success) {
      mediaState.isMicrophoneOpened.value = true;
    }
    return result;
  }

  Future<TUIActionCallback> unMuteLocalAudio() async {
    var result = await liveService.unMuteLocalAudio();
    if (result.code == TUIError.success) {
      mediaState.isMicrophoneMuted.value = false;
    }
    return result;
  }

  Future<TUIActionCallback> muteLocalAudio() async {
    var result = await liveService.muteLocalAudio();
    if (result.code == TUIError.success) {
      mediaState.isMicrophoneMuted.value = true;
    }
    return result;
  }

  void setLocalVideoView(int viewId) {
    liveService.setLocalVideoView(viewId);
  }

  void setRemoteVideoView(String userId, TUIVideoStreamType type, int viewId) {
    liveService.setRemoteVideoView(userId, type, viewId);
  }

  void startPlayRemoteVideo(String userId, TUIVideoStreamType type, TUIPlayCallback? playCallback) {
    liveService.startPlayRemoteVideo(userId, type, playCallback);
  }

  void stopPlayRemoteVideo(String userId, TUIVideoStreamType type) {
    liveService.stopPlayRemoteVideo(userId, type);
  }

  Future<TUIActionCallback> openLocalCamera() async {
    var cameraPermission = await Permission.camera.request();

    if (!cameraPermission.isGranted) {
      LiveKitLogger.error('requestMicrophonePermissions:[onDenied]');
      return TUIActionCallback(code: TUIError.errPermissionDenied, message: "camera permission denied");
    }

    var result = await liveService.openLocalCamera(mediaState.isFrontCamera.value, mediaState.videoQuality.value);
    if (result.code == TUIError.success) {
      _initLivingConfig();
      mediaState.isCameraOpened.value = true;
    }
    return result;
  }

  void closeLocalCamera() {
    liveService.closeLocalCamera();
    mediaState.isCameraOpened.value = false;
  }

  void switchCamera() async {
    bool isFrontCamera = mediaState.isFrontCamera.value;
    await liveService.switchCamera(!isFrontCamera);
    mediaState.isFrontCamera.value = !isFrontCamera;
  }

  void setCameraMirror() {
    bool isMirror = mediaState.isMirror.value;
    liveService.setCameraMirror(!isMirror);
    mediaState.isMirror.value = !isMirror;
  }

  void updateVideoQuality(TUIVideoQuality quality) {
    liveService.updateVideoQuality(quality);
    mediaState.videoQuality.value = quality;
  }

  void updateAudioQuality(TUIAudioQuality quality) {
    liveService.updateAudioQuality(quality);
    mediaState.audioQuality.value = quality;
  }

  void setBeautyLevel(int level) async {
    await liveService.setBeautyLevel(level);
    beautyState.smoothLevel.value = level;
  }

  void setWhitenessLevel(int level) async {
    await liveService.setWhitenessLevel(level);
    beautyState.whitenessLevel.value = level;
  }

  void setRuddyLevel(int level) async {
    await liveService.setRuddyLevel(level);
    beautyState.ruddyLevel.value = level;
  }

  void closeBeauty() {
    beautyState.smoothLevel.value = 0;
    liveService.setBeautyLevel(0);
    beautyState.whitenessLevel.value = 0;
    liveService.setWhitenessLevel(0);
    beautyState.ruddyLevel.value = 0;
    liveService.setRuddyLevel(0);
  }

  void _initLivingConfig() {
    liveService.enableGravitySensor(true);
    liveService.setVideoResolutionMode(TUIVideoStreamType.cameraStream, TUIResolutionMode.portrait);
    liveService.setBeautyStyle(0);
    updateVideoQuality(mediaState.videoQuality.value);
    setBeautyLevel(beautyState.smoothLevel.value);
    setWhitenessLevel(beautyState.whitenessLevel.value);
    setRuddyLevel(beautyState.ruddyLevel.value);
    updateAudioQuality(mediaState.audioQuality.value);
  }

  void _onLinkStateChanged() async {
    LiveKitLogger.info("$tag onUserStateChanged status:${viewState.linkStatus.value}.");
    if (viewState.linkStatus.value == LinkStatus.linking) {
      if (viewState.autoOpenCameraOnSeated.value) {
        LiveKitLogger.info("$tag requestPermissions:[Camera, Microphone]}");
        final openMicrophoneResult = await openLocalMicrophone();
        if (TUIError.success != openMicrophoneResult.code) {
          LiveKitLogger.error("$tag openLocalMicrophone:[message: ${openMicrophoneResult.message}]");
        }
        final openCameraResult = await openLocalCamera();
        if (TUIError.success == openCameraResult.code) {
          LiveKitLogger.error("$tag openLocalCamera:[message: ${openMicrophoneResult.message}]");
        }
      } else {
        await openLocalMicrophone();
      }
    }
  }
}
